torch.fft: Multi-dimensional transforms

peterbell10 opened the pull request:
Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45164 torch.fft: Two dimensional FFT functions
* #45163 codegen: Improve array default handing
* #44877 torch.fft helper functions
* #44876 Deprecate old fft functions
* **#44550 torch.fft: Multi-dimensional transforms**

Differential Revision: [D23846032](https://our.internmc.facebook.com/intern/diff/D23846032)

peterbell10 committed:
torch.fft: Multi-dimensional transforms

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.




[ghstack-poisoned]

mruberry commented on aten/src/ATen/WrapDimUtils.h:
```diff
@@ -30,14 +30,15 @@ static inline int64_t maybe_wrap_dim(int64_t dim, const std::vector<std::vector<
-// wrap each of dims basing on dim_post_expr
-static inline void maybe_wrap_dims(std::vector<int64_t>& dims, int64_t dim_post_expr) {
+static inline void maybe_wrap_dims_n(int64_t * dims, int64_t ndims, int64_t dim_post_expr) {
+  // wrap each of dims basing on dim_post_expr
```
Can we improve this comment? I'm not sure what "basing on dim_post_expr" means.

peterbell10 replied:
Reworded to "taking dim_post_expr as the true number of dimensions".

mruberry commented on aten/src/ATen/WrapDimUtils.h:
```diff
@@ -30,14 +30,15 @@ static inline int64_t maybe_wrap_dim(int64_t dim, const std::vector<std::vector<
   int64_t min = -dim_post_expr;
   int64_t max = dim_post_expr - 1;
-  for (auto& dim : dims) {
+  for (int64_t i = 0; i < ndims; ++i) {
```
Why make this change?

peterbell10 replied:
To support passing `DimVector` in place of `std::vector`.

mruberry commented on aten/src/ATen/WrapDimUtils.h:
```diff
@@ -30,14 +30,15 @@ static inline int64_t maybe_wrap_dim(int64_t dim, const std::vector<std::vector<
 
-// wrap each of dims basing on dim_post_expr
-static inline void maybe_wrap_dims(std::vector<int64_t>& dims, int64_t dim_post_expr) {
+static inline void maybe_wrap_dims_n(int64_t * dims, int64_t ndims, int64_t dim_post_expr) {
```
`int64_t*`

mruberry commented on aten/src/ATen/WrapDimUtils.h:
```diff
@@ -47,6 +48,11 @@ static inline void maybe_wrap_dims(std::vector<int64_t>& dims, int64_t dim_post_
   }
 }
 
+template <typename Container>
```
Add a brief comment before this function explaining its purpose/effect

mruberry commented on torch/fft/__init__.py:
```diff
@@ -87,6 +87,98 @@
+
+Computes the N dimensional discrete Fourier transform of :attr:`input`.
+
+Note:
```
".. note::", right?

peterbell10 replied:
`Note` is one of sections supported by napoleon:
https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html#docstring-sections

mruberry commented on torch/fft/__init__.py:
```diff
@@ -87,6 +87,98 @@
+    The Fourier domain representation of any real signal satisfies the
+    Hermitian property: ``X[i_1, ..., i_n] = conj(X[-i_1, ..., -i_n])``. This
+    function always returns all positive and negative frequency terms, even
+    though for real inputs, half of these values are redundant.
```
"even though, for real inputs, half of these values are redundant."

mruberry commented on torch/fft/__init__.py:
```diff
@@ -87,6 +87,98 @@
+    Hermitian property: ``X[i_1, ..., i_n] = conj(X[-i_1, ..., -i_n])``. This
+    function always returns all positive and negative frequency terms, even
+    though for real inputs, half of these values are redundant.
+    :func:`~torch.fft.rfftn` returns the more efficient "half-complex"
```
"efficient" is vague. What about "returns the more compact..."

I know we discussed this briefly on the last PR, but is "half-complex" really the term of art for this?

mruberry replied:
If there's not a good term of art, do we need to name this?

"returns a more compact representation containing only the positive frequencies of the last dimension."

peterbell10 replied:
This wording comes from the `rfft` docs. I must have missed it when I changed "half-complex" to "one-sided" elsewhere. I've updated both PRs to use "one-sided".

mruberry commented on torch/fft/__init__.py:
```diff
@@ -87,6 +87,98 @@
+    though for real inputs, half of these values are redundant.
+    :func:`~torch.fft.rfftn` returns the more efficient "half-complex"
+    representation where only the positive frequencies of the last dimension
+    are stored explicitly.
```
"are stored explicitly" is vague. Stating that rfftn returns those values is sufficient, I think.

peterbell10 replied:
Changed to "are returned".

mruberry commented on torch/fft/__init__.py:
```diff
@@ -87,6 +87,98 @@
+    s (Tuple[int], optional): Signal size in the transformed dimensions. If
+        given, the input will either be zero-padded or trimmed to this length
+        before computing the FFT.
+    dim (Tuple[int], optional): Dimensions to be Fourier transformed.
```
"Dimensions to be transformed" is sufficient. If s and dim are given, do they have to be the same size?

In the NumPy version of this function repeated dims are allowed, although the NumPy documentation isn't very clear about what happens when they're provided. What happens if dims are repeated here?

peterbell10 replied:
>In the NumPy version of this function repeated dims are allowed, although the NumPy documentation isn't very clear about what happens when they're provided. What happens if dims are repeated here?

In NumPy, if you give the same dimension multiple times it will FFT that dimension twice which I don't think is a meaningful transformation. This PR follows the SciPy interface which checks for unique dimensions. That was done because we though it was more likely to be user error than a desirable feature.

Supporting repeated transforms would be possible but would require more complication to support. It also wouldn't be any much efficient than just calling `fftn` multiple times.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -87,6 +87,98 @@
+        * ``"backward"`` - no normalization
+        * ``"ortho"`` - normalize by ``1/sqrt(n)`` (making the FFT orthonormal)
+
+        Where ``n = prod(s)`` is the logical FFT size.
```
"s" may not be specified, so `n = prod(s)` will have to change.

peterbell10 replied:
How about if I add to the docs:
> Default: `s = [input.size(d) for d in dim]`

mruberry replied:
OK.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -87,6 +87,98 @@
+    >>> x = torch.rand(10, 10, dtype=torch.complex64)
+    >>> fftn = torch.fft.fftn(t)
+
+    The Fourier transform is separable, so :func:`~torch.fft.fftn` here is
```
Technically it would be the "discrete" or "fast" Fourier transform in this case

mruberry commented on torch/fft/__init__.py:
```diff
@@ -87,6 +87,98 @@
+    >>> x = torch.rand(10, 10, dtype=torch.complex64)
+    >>> ifftn = torch.fft.ifftn(t)
+
+    The Fourier transform is separable, so :func:`~torch.fft.ifftn` here is
```
Won't the normalization be different between two ffts and a single fftn, though?

peterbell10 replied:
Consider an N by M transform:
- `fftn` normalizes by `N*M` directly.
- 1st `fft` normalizes by `N`, 2nd `fft` normalizes by `M`. Overall result is an `N*M` normalization.

peterbell10 replied:
You can also see that NumPy implements `fftn` exactly as multiple calls to `fft` without any changes to the normalization behavior:
https://github.com/numpy/numpy/blob/3b294521a7394c2195e6b2980555c6b27258bac9/numpy/fft/_pocketfft.py#L699-L706

mruberry commented on torch/fft/__init__.py:
```diff
@@ -199,6 +291,133 @@
     tensor([0.0000, 1.0000, 2.0000, 3.0000, 4.0000])
 """)
 
+rfftn = _add_docstr(_fft.fft_rfftn, r"""
```
Some of the same questions as above here, too.

mruberry commented on torch/fft/__init__.py:
```diff
@@ -199,6 +291,133 @@
+    >>> rfftn.size()
+    torch.Size([10, 6])
+
+    Compared against the full output from :func:`~torch.fft.fftn`, we have all
```
Cool example

mruberry commented on torch/fft/__init__.py:
```diff
@@ -199,6 +291,133 @@
+    the original data, as given by :attr:`s`. This is because each input shape
+    could correspond to either an odd or even length signal. By default, the
+    signal is assumed to be even length and odd signals will not round-trip
+    properly. So, it is recommended to always pass the signal length :attr:`s`:
```
"signal length(s)"?

peterbell10 replied:
changed to "signal shape"

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -157,11 +158,13 @@ def test_fft_round_trip(self, device, dtype):
         match = r"Invalid number of data points \([-\d]*\) specified"
-        fft_functions = [torch.fft.fft, torch.fft.ifft, torch.fft.hfft,
-                         torch.fft.irfft]
+        fft_functions = [torch.fft.fft, torch.fft.fftn,
```
Not in this PR, but in a follow-up it'd be nice to see the fft functions added to the OpInfo database in common_methods_invocations.py. They could have their own base class derived from OpInfo. Registering them as OpInfos would make these tests a little nicer, and we'll be moving all `method_tests` registrations to OpInfos in the future, anyway.

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -203,6 +203,103 @@ Tensor fft_c2c(Tensor input, c10::optional<int64_t> n_opt,
 
+struct ShapeAndDims{
+  DimVector shape, dim;
+};
```
newline

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -203,6 +203,103 @@ Tensor fft_c2c(Tensor input, c10::optional<int64_t> n_opt,
   return out;
 }
 
+struct ShapeAndDims{
```
brief comment explaining this struct

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -203,6 +203,103 @@ Tensor fft_c2c(Tensor input, c10::optional<int64_t> n_opt,
+struct ShapeAndDims{
+  DimVector shape, dim;
+};
+ShapeAndDims canonicalize_fft_shape_and_dim_args(
```
Add a brief comment explaining this function and its intended use

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -203,6 +203,103 @@ Tensor fft_c2c(Tensor input, c10::optional<int64_t> n_opt,
+    TORCH_CHECK(!dim || dim->size() == shape->size(),
+                "When given, dim and shape arguments must have the same length");
+    TORCH_CHECK(shape->size() <= input_dim,
+                "Shape requires more dims than are present");
```
Maybe something like: "Got X shape values but input tensor only has Y dimensions."?

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -203,6 +203,103 @@ Tensor fft_c2c(Tensor input, c10::optional<int64_t> n_opt,
+    TORCH_CHECK(shape->size() <= input_dim,
+                "Shape requires more dims than are present");
+    const int64_t transform_ndim = shape->size();
+    if (!dim) {
```
Add a comment for this section, too (I think the doc may also need an update to better describe this behavior). Maybe something like, "If shapes are specified but dims aren't, the last s dims of the input are transformed."?

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -203,6 +203,103 @@ Tensor fft_c2c(Tensor input, c10::optional<int64_t> n_opt,
+      const auto n = (*shape)[i];
+      ret.shape[i] = n == -1 ? input_sizes[ret.dim[i]] : n;
+    }
+  } else if (!dim) {
```
Add a comment for each branch of this conditional to make it easier to follow:

- has shape, maybe has dim
- no shape, no dim
- no shape, dim

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -203,6 +203,103 @@ Tensor fft_c2c(Tensor input, c10::optional<int64_t> n_opt,
+      std::iota(ret.dim.begin(), ret.dim.end(), input_dim - transform_ndim);
+    }
+
+    ret.shape.resize(transform_ndim);
```
Add a brief comment explaining this block.

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -203,6 +203,103 @@ Tensor fft_c2c(Tensor input, c10::optional<int64_t> n_opt,
+  return ret;
+}
+
+// Complex to complex nd fft
```
"nd" -> "n-dimensional"

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -203,6 +203,103 @@ Tensor fft_c2c(Tensor input, c10::optional<int64_t> n_opt,
+  const auto norm = norm_from_string(norm_str, forward);
+  constexpr int64_t max_signal_ndim = 3;
+
+  // Transform n dimensions, up to 3 at a time
```
Would you add a line here explaining why there's a 3 dim at a time limit?

mruberry replied:
(or just above the definition of `max_signal_ndim`)

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -203,6 +203,103 @@ Tensor fft_c2c(Tensor input, c10::optional<int64_t> n_opt,
+
+  // Transform n dimensions, up to 3 at a time
+  // TODO: rewrite _fft_with_size to transform more than 3 dimensions at once.
+  for (int64_t i = 0; i < transform_ndim; i += max_signal_ndim) {
```
This goes back to my question about how normalization works. Is each dim normalized separately? If not, won't doing 3 dims at a time not work because the call to `_fft` won't be able to compute `n` properly?

mruberry commented on aten/src/ATen/native/SpectralOps.cpp:
```diff
@@ -240,6 +337,63 @@ Tensor fft_ihfft(const Tensor& self, c10::optional<int64_t> n, int64_t dim,
+    if (s.has_value()) {
+      return desc.shape.back();
+    } else {
+      return c10::nullopt;
```
This needs a comment

mruberry commented:
Hey @peterbell10! Looks good as always. I have some questions and made some comment/doc requests, nothing too significant. 

Looking forward to getting multi-dim ffts into PyTorch!

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.




[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.




[ghstack-poisoned]

mruberry commented:
Hey @peterbell10, this is looking great. Let me know when I should review it again and if there's anything I can help with. As mentioned on the next PR: 1.7's branch cut is 8 days away, so I'd like to get this in within the next week.

peterbell10 commented:
@mruberry this is ready for another look.

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.




[ghstack-poisoned]

mruberry reviewed (approved):
Awesome work as usual, @peterbell10!

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.




[ghstack-poisoned]

peterbell10 commented:
Fixed a few small doc errors which weren't updated properly after copy-pasting.

mruberry commented on test/test_spectral_ops.py:
```diff
@@ -340,6 +348,164 @@ def test_fn(x):
+
+    @skipCPUIfNoMkl
+    @skipCUDAIfRocm
+    @unittest.skipIf(not TEST_NUMPY, 'NumPy not found')
```
You may need `@onlyOnCPUAndCUDA` here and on another test below. Would you rebase to test the XLA build?

peterbell10 replied:
Noticed some of the `fft` tests were missing this as well. Added and rebased on `viable/strict`. 

@mruberry

mruberry replied:
Thanks. XLA build looks good.

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.


Differential Revision: [D23846032](https://our.internmc.facebook.com/intern/diff/D23846032)

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.


Differential Revision: [D23846032](https://our.internmc.facebook.com/intern/diff/D23846032)

[ghstack-poisoned]

peterbell10 committed:
Update on "torch.fft: Multi-dimensional transforms"

Part of the `torch.fft` work (gh-42175).
This adds n-dimensional transforms: `fftn`, `ifftn`, `rfftn` and `irfftn`.

This is aiming for correctness first, with the implementation on top of the existing `_fft_with_size` restrictions. I plan to follow up later with a more efficient rewrite that makes `_fft_with_size` work with arbitrary numbers of dimensions.


Differential Revision: [D23846032](https://our.internmc.facebook.com/intern/diff/D23846032)

[ghstack-poisoned]
